File: src/rust-nfv/src/nf1.rs
use crate::packet::{Packet, IPV4_TTL_OFFSET};

pub struct Nf1DecrementTtl {}

impl Nf1DecrementTtl {
  pub fn new() -> Self {
    Self {}
  }
}

impl crate::nfv::NetworkFunction for Nf1DecrementTtl {
    fn process_frames(&mut self, packets: &mut[Packet]) {
        for pkt in packets.iter_mut() {
            let ip_header = pkt.ip_header();
            ip_header[IPV4_TTL_OFFSET] -= 1;
        }
    }
}

File: src/libnfv/nf1.cpp
#include "nf1.hpp"

#include <tuple>

#include <rte_ether.h>
#include <rte_ip.h>

#include "nfv.hpp"
#include "packettool.hpp"

void NF1DecrementTtl::_process_frames(const std::span<rte_ether_hdr*> packets, int buffer_id) {
  for (auto&& packet : packets) {
    // Get packet header.
    rte_ipv4_hdr* ipv4_hdr = get_ipv4_hdr(packet);
    if (ipv4_hdr == nullptr) {
      continue;
    }
    // Decrement TLL
    ipv4_hdr->time_to_live--;
  }
}

void NF1DecrementTtl::report() {
}
